pkg/services/vimmachine.go

```diff
@@ -446,25 +450,66 @@ func (v *VimMachineService) generateOverrideFunc(ctx context.Context, vimMachine
 // The substitution is done based on the order in which the network devices have been defined.
 //
 // In case there are more network definitions than the number of network devices specified, the definitions are appended to the list.
-func overrideNetworkDeviceSpecs(deviceSpecs []infrav1.NetworkDeviceSpec, networks []string) []infrav1.NetworkDeviceSpec {
+func overrideNetworkDeviceSpecs[T any](deviceSpecs []infrav1.NetworkDeviceSpec, networks []T, mergeFunc func(device *infrav1.NetworkDeviceSpec, network T)) []infrav1.NetworkDeviceSpec {
 	index, length := 0, len(networks)
 
-	devices := make([]infrav1.NetworkDeviceSpec, 0, max(length, len(deviceSpecs)))
+	devices := make([]infrav1.NetworkDeviceSpec, 0, integer.IntMax(length, len(deviceSpecs)))
 	// override the networks on the VM spec with placement constraint network definitions
 	for i := range deviceSpecs {
 		vmNetworkDeviceSpec := deviceSpecs[i]
 		if i < length {
 			index++
-			vmNetworkDeviceSpec.NetworkName = networks[i]
+			mergeFunc(&vmNetworkDeviceSpec, networks[i])
 		}
 		devices = append(devices, vmNetworkDeviceSpec)
 	}
 	// append the remaining network definitions to the VM spec
 	for ; index < length; index++ {
-		devices = append(devices, infrav1.NetworkDeviceSpec{
-			NetworkName: networks[index],
-		})
+		device := infrav1.NetworkDeviceSpec{}
+		mergeFunc(&device, networks[index])
+		devices = append(devices, device)
 	}
 
 	return devices
 }
+
+func mergeFailureDomainNetworkName(device *infrav1.NetworkDeviceSpec, network string) {
+	device.NetworkName = network
+}
+
+func mergeFailureDomainNetSpecToNetworkDeviceSpec(device *infrav1.NetworkDeviceSpec, fd infrav1.NetworkConfiguration) {
```

Comment:
```suggestion
func mergeNetworkConfigurationInNetworkDeviceSpec(device *infrav1.NetworkDeviceSpec, fd infrav1.NetworkConfiguration) {
```

more of a question whether having the types in the function makes more sense?
and on that note should `fd` be replaced with something like `nc`?

Reply:
Fixed names as suggested. If you mean types in function name, i don't see other choice due to lack of overloading in go.